tests/test_upload.py

```diff
@@ -444,3 +447,73 @@ def test_upload_registration_fail(rse, scope, upload_client_registration_fail, f
             'did_name': fn,
             'guid': generate_uuid()
         }])
+
+
+@pytest.mark.parametrize("structure,failure", [
+    ({"container": ["file1", "file2", {"dataset": ["file3", "file4"]}]}, True),  # Cannot have both files and datasets in a container
+    ({
+        "container": {
+            "dataset": ["file1", "file2"]
+            },
+        "container2": {
+            "dataset": ["file1", {"dataset2": ["file1"]}]
+        }
+    }, True),  # One valid container, another invalid container
+    ({
+        "container": {"container2": {"dataset": {"dataset2": ["file1", {"dataset3": ["file2"]}]}}},
+    }, True),  # Same issue but further down the tree
+    ({"container": {"dataset": ["file1"], "dataset2": ["file2"]}}, False),  # One valid container
+    ({
+        "container": {"dataset": ["file1"], "dataset2": ["file2"]},
+        "container2": {"dataset": ["file1"], "dataset2": ["file2"]}
+    }, False)  # Two valid containers
+])
```

Comment:
Can you use the `ids=` feature to move the comments here into ids?

e.g.

https://github.com/rucio/rucio/blob/33f1aa716f7032bf4b1e9888532332bb643a352b/tests/rucio/common/test_types.py#L23-L37

This way, if one of these test cases fails, the test logs show up the ID of the failing case, which would make it simpler to debug it.